### PR TITLE
cmake support win32 ,and support generate&use prebuilt libs

### PR DESCRIPTION
**Pull request by drelaptop**

### CODE REVIEW

**Comment by drelaptop** [comment]
@newnon when build cpp-tests project __Release__ , using Android Studio native build by cmake, link failed, this is the detail info.
```
[840/840] Linking CXX shared library ..\..\..\..\build\intermediates\cmake\release\obj\armeabi-v7a\libcpp_tests.so
  FAILED: cmd.exe /C "cd . && C:\projects\cocos2d-x\..\android-ndk-r16b\toolchains\llvm\prebuilt\windows-x86_64\bin\clang++.exe  --target=armv7-none-linux-androideabi 
--gcc-toolchain=C:/projects/cocos2d-x/../android-ndk-r16b/toolchains/arm-linux-androideabi-4.9/prebuilt/windows-x86_64 --sysroot=C:/projects/cocos2d-x/../android-ndk-r16b/sysroot -fPIC -isystem 
C:/projects/cocos2d-x/../android-ndk-r16b/sysroot/usr/include/arm-linux-androideabi -D__ANDROID_API__=14 -g -DANDROID -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -march=armv7-a 
-mfloat-abi=softfp -mfpu=vfpv3-d16 -fno-integrated-as -mthumb -mfpu=neon -Wa,--noexecstack -Wformat -Werror=format-security -std=c++11 -frt
ti -fexceptions -std=c++11 -Wno-deprecated-declarations -Wno-reorder -Wno-invalid-offsetof -fPIC -fsigned-char -fexceptions -Os -DNDEBUG  -Wl,--exclude-libs,libgcc.a -Wl,--exclude-libs,libatomic.a --sysroot 
C:/projects/cocos2d-x/../android-ndk-r16b/platforms/android-14/arch-arm -Wl,--build-id -Wl,--warn-shared-textrel -Wl,--fatal-warnings -Wl,--fix-cortex-a8 -Wl,--exclude-libs,libunwind.a 
-LC:/projects/cocos2d-x/../android-ndk-r16b/sources/cxx-stl/llvm-libc++/libs/armeabi-v7a -Wl,--no-undefined -Wl,-z,noexecstack -Qunused-arguments -Wl,-z,relro -Wl,-z,now -shared -Wl,-soname,libcpp_tests.so -o 
..\..\..\..\build\intermediates\cmake\release\obj\armeabi-v7a\libcpp_tests.so @CMakeFiles/cpp_tests.rsp  -latomic && cmd.exe /C "cd /D 
C:\projects\cocos2d-x\tests\cpp-tests\proj.android\app\.externalNativeBuild\cmake\release\armeabi-v7a && 
C:\projects\android-ndk-r16b\toolchains\arm-linux-androideabi-4.9\prebuilt\windows-x86_64\bin\arm-linux-androideabi-strip.exe libcpp_tests.so""
```
the CI test this, https://ci.appveyor.com/project/minggo/cocos2d-x/build/1.0.768/job/snsbg2w3ajqyogjt, when I test in my Android Studio, set mode to Release, link failed too, would you like check it and fix it?

**Comment by newnon** [comment]
I will try to build it on Mac Os may be it is some kind windows problem? (like max path length 260 symbols)

**Comment by drelaptop** [comment]
thanks, but I think it's not a windows problem, because my Android Studio is on macOS.

**Comment by drelaptop** [comment]
```
cmake -H"..\cocos2d-x\" -Bmsvc_build -G"Visual Studio 15 2017" -DCMAKE_BUILD_TYPE=Debug
cmake -H"..\cocos2d-x\" -Bmsvc_build -G"Visual Studio 15 2017" -DCMAKE_BUILD_TYPE=Release
```
For now, we can use command like above to generate Visual Studio project, and then open the `Cocos2d-x.sln` in _msvc_build_ folder, every project in `Cocos2d-x.sln` works well. I test this in `Visual Studio 2017` on Windows 10.

__Attantion__
1. lua template project not add in `Cocos2d-x.sln`, it need to be done next
1. Debug and Release mode all tested pass, please remember that `-DCMAKE_BUILD_TYPE=Debug` not `DEBUG` or `debug`
1. compile options and link options might not perfect, If you have suggestions when you test, please tell me, I will quite appreciate

**Comment by crazyhappygame** [comment]
@drelaptop Nice windows android build passed. When do you integrate changes?

Could you check travis build? It seems like something is broken with android/linux build e.g.: https://travis-ci.org/cocos2d/cocos2d-x/jobs/346707241. Build is marked as green but it took only 6 minutes (no cpp file compiling). Also mac/ios build fails on download-deps.py it seems like some infrastructure issue.

**Comment by drelaptop** [comment]
@crazyhappygame I prepare to do `cmake support prebuilt libs` in this PR, so it would be merged after prebuilt works done. this work might spend a week.

thanks for mention it, I will try to fix cmake build error it after prebuilt works done. As for this error:`URLError: <urlopen error [SSL: TLSV1_ALERT_PROTOCOL_VERSION] tlsv1 alert protocol version (_ssl.c:590)>
` I don't know why

**Comment by crazyhappygame** [comment]
I added "cocos new" tests in https://github.com/cocos2d/cocos2d-x/pull/18703 .
I found that linux build fails for new project. Could you check this:
https://travis-ci.org/cocos2d/cocos2d-x/builds/347005988?utm_source=github_status&utm_medium=notification

**Comment by drelaptop** [comment]
generate and use prebuilt is simple, you only need add 2 or 3 cmake variable.
```
add "-DGEN_COCOS_PREBUILT=ON" and "-DUSE_COCOS_PREBUILT=OFF" // will add target prebuilt, build the target will generate prebuilt libs
add "-DGEN_COCOS_PREBUILT=OFF" and "-DUSE_COCOS_PREBUILT=ON"  //not include libs target and prebuilt target, directly use the prebuilt libs in prebuilt directory
//add "COCOS_PREBUILT_ROOT" to set prebuilt libs directory, the default prebuilt directory is `cocos2d-x-root/prebuilt`, cmake get the path by `$ENV{COCOS_TEMPLATES_ROOT}/../prebuilt`, but when using cmake in android studio, it can't get environment variable `COCOS_TEMPLATES_ROOT`, so you should set the prebuilt directory manually. the corresponding cmake variable is `COCOS_PREBUILT_ROOT`
```
Attention: the default prebuilt behavior is `GEN_COCOS_PREBUILT` and `USE_COCOS_PREBUILT` both `OFF`, and don't set both cmake variable `ON` please

**Comment by crazyhappygame** [comment]
@drelaptop Great job. Could you write example how can use pre-builds for 3 new create apps (3 games, using 1 prebuild of cocos2d-x):
```
$ cd cocos2d-x
$ python tools/cocos2d-console/bin/cocos.py new -l cpp -p my.pack.app1 test_app1
$ python tools/cocos2d-console/bin/cocos.py new -l cpp -p my.pack.app2 test_app2
$ python tools/cocos2d-console/bin/cocos.py new -l cpp -p my.pack.app3 test_app3
$. ....
```

**Comment by drelaptop** [comment]
same options to use prebuilt by CMake in every platforms, for example macOS platforms and only pre-build C++ related libs

add `-DGEN_COCOS_PREBUILT=ON` to generate prebuilt libs, it will be saved to cocos2d-x/prebuilt directory default

> add `COCOS_PREBUILT_ROOT` to set prebuilt libs directory manually

```sh
cocos new -l cpp -p my.pack.app1 test_app1
mkdir app1_build && cd app1_build
cmake ../test_app1 -DGEN_COCOS_PREBUILT=ON
make prebuilt
```
close option `GEN_COCOS_PREBUILT` and open `USE_COCOS_PREBUILT` to use prebuilt in the same project
```sh
cmake ../test_app1 -DGEN_COCOS_PREBUILT=OFF -DUSE_COCOS_PREBUILT=ON
make TemplateCpp
open bin/TemplateCpp.app
```

add `-DUSE_COCOS_PREBUILT=ON` to use prebuilt generated before, default value of  `GEN_COCOS_PREBUILT` and `USE_COCOS_PREBUILT` both are `OFF`.
> please set  `COCOS_PREBUILT_ROOT` if the prebuilt libs don't located at default directory 

```sh
cocos new -l cpp -p my.pack.app2 test_app2
mkdir app2_build && cd app2_build
cmake ../test_app2 -DUSE_COCOS_PREBUILT=ON
make TemplateCpp
open bin/TemplateCpp.app
```
> any other cpp project can use prebuilt in this way

Other Option:
add `-GXcode` to generate Xcode project
```sh
cmake ../test_app2 -DUSE_COCOS_PREBUILT=ON -GXcode
open TemplateCpp.xcodeproj
```
add `-DCMAKE_BUILD_TYPE=Release` or `Debug` to generate Release or Debug project, the default is `Debug`

**Comment by Anywhere-Music-Player** [comment]
USE_SIMULATOR - it's not needed.

**Comment by drelaptop** [comment]
@KAMIKAZEUA what's the meaning of `USE_SIMULATOR `, I didn't use this option

**Comment by drelaptop** [comment]
this is the simple cmake using docs for this feature, it include common options and build example.

https://github.com/drelaptop/cocos2d-x/blob/v3-cmake-win32/cmake/README.md

please review and merge @minggo

File: cmake/Modules/CocosBuildHelpers.cmake [hunk 1]
```diff
@@ -1,28 +1,23 @@
...
-    COMMENT "${TARGET_NAME} POST_BUILD ..."
-    )
-endfunction()
-
 # lib_name eg cocos2d/cocos2djs
-macro(cocos_find_static_libs lib_name)
-  # only search COCOS_PREBUILT_LIBS_PATH
-  MESSAGE( STATUS "cocos static library path: ${COCOS_PREBUILT_LIBS_PATH}")
-  FIND_LIBRARY(LIB_FOUND ${lib_name} PATHS ${COCOS_PREBUILT_LIBS_PATH} DOC "using cocos static library: lib${lib_name}.a" NO_DEFAULT_PATH NO_CMAKE_FIND_ROOT_PATH)
+function(cocos_find_prebuilt lib_target lib_ret)
```
Comment (minggo) [inline cmake/Modules/CocosBuildHelpers.cmake]: `cocos_find_prebuilt lib_target lib_ret` this function name is strange, what's the meaning? Why not rename to `cocos_find_prebuilt lib`?

Comment (minggo) [reply cmake/Modules/CocosBuildHelpers.cmake]: Oh, `lib_ret` and lib_target  are parameters. But what's the meaning of them?

File: cmake/Modules/CocosBuildHelpers.cmake [hunk 2]
```diff
@@ -94,14 +204,104 @@ function(cocos_mark_code_files cocos_target)
   
 endfunction()
 
+# APP_SRC app needed all src files
+# DEPEND_COMMON_LIBS the app needed libs for all platforms
+# DEPEND_ANDROID_LIBS the app needed libs only for android platform
+macro(cocos_mark_app app_name)
```
Comment (minggo) [inline cmake/Modules/CocosBuildHelpers.cmake]: what does `cocos_mark_app ` do? I can not understand it from the macro name. And is it better to use capital characters for macro?

File: cmake/Modules/SelectModule.cmake [hunk 3]
```diff
@@ -28,20 +28,20 @@ macro (SelectModule)
...
   option(BUILD_EDITOR_COCOSBUILDER "Build editor support for cocosbuilder" ON)
   option(BUILD_BOX2D "Build box2d external without using it for physics library" OFF)
-  option(BUILD_LUA_LIBS "Build lua libraries" ${BUILD_LUA_LIBS_DEFAULT})
-  option(BUILD_JS_LIBS "Build js libraries" ${BUILD_JS_LIBS_DEFAULT})
-  # need review
-  option(USE_EXTERNAL_PREBUILT_LIBS "Use prebuilt libraries in external directory" ${USE_PREBUILT_LIBS_DEFAULT})
-  option(USE_COCOS_PREBUILT_LIBS "use cocos static libraries, return error if not find" ${USE_COCOS_PREBUILT_LIBS_DEFAULT})
-  option(USE_SOURCES_EXTERNAL "Use sources in external directory (automatically ON when USE_EXTERNAL_PREBUILT_LIBS is ON)" ${USE_SOURCES_EXTERNAL_DEFAULT})
+  option(USE_LUACOCOS2D "Build lua libraries" ${USE_LUACOCOS2D_DEFAULT})
+  option(USE_JSCOCOS2D "Build js libraries" ${USE_JSCOCOS2D_DEFAULT})
```
Comment (minggo) [inline cmake/Modules/SelectModule.cmake]: I think new names are not good as before.

**Comment by Anywhere-Music-Player** [comment]
I guess it’s like here https://github.com/cocos2d/cocos2d-console/issues/413

**Comment by drelaptop** [comment]
you can new a cpp project, and then add option `-DGEN_COCOS_PREBUILT=ON`, you will get cpp prebuilt libs, not include lua and js lib. this is friendly to use. as the answer above of  

> Could you write example how can use pre-builds for 3 new create apps (3 games, using 1 prebuild of cocos2d-x):

lua and js language are same too. @KAMIKAZEUA

File: cmake/Modules/CocosBuildHelpers.cmake (hunk 1 above)
Comment (drelaptop) [reply cmake/Modules/CocosBuildHelpers.cmake]: Should I add detail comment for every macro or function parameters? 

In this macro `lib_target ` is a lib name, means which lib do you want to find in prebuilt directory. `lib_ret` is the found lib, it save the absolute path of lib. it's a function `lib_ret` like return value, for cmake's function need to set return value like this `set(${lib_ret} ${tmp_lib} PARENT_SCOPE)`

I think use function is a better choice than macro, it didn't have effect on outer space, we can keep focus on the input parameters and output value.

File: cmake/Modules/CocosBuildHelpers.cmake (hunk 2 above)
Comment (drelaptop) [reply cmake/Modules/CocosBuildHelpers.cmake]: input a name, and other application related informations, mark those as a cocos application which name is `app_name`.

this macro reduce much duplicate code, for those content is necessary for every application.  for example mark cpp_empty_test application.
```
# mark app complie info and libs info
cocos_mark_app(${APP_NAME} 
    APP_SRC "${all_code_files}"
    DEPEND_COMMON_LIBS "cocos2d"
    DEPEND_ANDROID_LIBS "cocos2d_android"
)
```

File: cmake/Modules/CocosBuildHelpers.cmake (hunk 1 above)
Comment (minggo) [reply cmake/Modules/CocosBuildHelpers.cmake]: I think it is better to have better names, and add some comments for it, for example

```
# find a prebuilt lib by `name` and save the result in `out`
function(cocos_find_prebuilt_lib_by_name name out)
```

File: cmake/Modules/SelectModule.cmake (hunk 3 above)
Comment (drelaptop) [reply cmake/Modules/SelectModule.cmake]: agree, keep the variable name as before is a better choice

File: cmake/Modules/CocosBuildHelpers.cmake (hunk 1 above)
Comment (drelaptop) [reply cmake/Modules/CocosBuildHelpers.cmake]: agree, I will do some improvements.

File: cmake/Modules/CocosBuildHelpers.cmake (hunk 2 above)
Comment (drelaptop) [reply cmake/Modules/CocosBuildHelpers.cmake]: meanwhile it's useful to hide the details of using compiled lib or prebuilt lib, it make the application CMakeFiles.txt more clear, like [cpp-template-default/CMakeLists.txt](https://github.com/drelaptop/cocos2d-x/blob/v3-cmake-win32/templates/cpp-template-default/CMakeLists.txt)

macro is quite like a function in cmake, I will try to switch this macro to a function, if I can't I change to use capital chars for macro, and lower-case for function name.

`cocos_mark_app` isn't a good name, I will change it to `cocos_build_app`, `mark` is used to mark res files

**Comment by crazyhappygame** [comment]
Thank you for answer. Once again great job.

Could you check your latest travis build? https://travis-ci.org/cocos2d/cocos2d-x/jobs/350245653 . BUILD_TARGET=android is green but there is no cpp files building... ( you can integrate this PR and fix travis BUILD_TARGET=android in next iteration)

**Comment by drelaptop** [comment]
@crazyhappygame  thanks for mention it, I want to this PR merged first, then solve the issue.

improve variable name, add comments for function finished, CI checks passed. please review and merge @minggo

**Pull request merged**